src/borg/crypto/key.py

```diff
@@ -432,13 +432,33 @@ def decrypt_key_file(self, data, passphrase):
         unpacker = get_limited_unpacker('key')
         unpacker.feed(data)
         data = unpacker.unpack()
-        enc_key = EncryptedKey(internal_dict=data)
-        assert enc_key.version == 1
-        assert enc_key.algorithm == 'sha256'
-        key = passphrase.kdf(enc_key.salt, enc_key.iterations, 32)
-        data = AES(key, b'\0'*16).decrypt(enc_key.data)
-        if hmac_sha256(key, data) == enc_key.hash:
+        encrypted_key = EncryptedKey(internal_dict=data)
+        if encrypted_key.version == 1:
+            return self.decrypt_key_file_v1(encrypted_key, passphrase)
+        elif encrypted_key.version == 2:
+            return self.decrypt_key_file_v2(encrypted_key, passphrase)
+        else:
+            raise Exception('Your borg key is stored in an unsupported format. Try using a newer version of borg')
+
+    def decrypt_key_file_v1(self, encrypted_key, passphrase):
+        assert encrypted_key.algorithm == 'sha256'
+        key = passphrase.kdf(encrypted_key.salt, encrypted_key.iterations, 32)
+        data = AES(key, b'\0'*16).decrypt(encrypted_key.data)
+        if hmac_sha256(key, data) == encrypted_key.hash:
             return data
+        return None
+
+    def decrypt_key_file_v2(self, encrypted_key, passphrase):
+        enc_key, mac_key = passphrase.argon2(
+            salt=encrypted_key.salt,
+            time_cost=encrypted_key.argon2_time_cost,
+            memory_cost=encrypted_key.argon2_memory_cost,
+            parallelism=encrypted_key.argon2_parallelism,
+            type=encrypted_key.argon2_type,
+        )
+        if not hmac.compare_digest(encrypted_key.hash, hmac_sha256(mac_key, encrypted_key.data)):
+            return None
+        return AES(enc_key, b'\0'*16).decrypt(encrypted_key.data)
```

Comment:
Just had the idea if we maybe want to use chacha20-poly1305 here instead of self-made authenticated encryption.

Reply:
otoh, openbsd with libressl does not have it, so guess we better don't. :-(

Reply:
`borg.crypto.low_level.AES256_CTR_HMAC_SHA256` could be used.

that way we maybe could get rid of the legacy `AES` class at some time in the future.

Reply:
I'll switch from `borg.crypto.low_level.AES` to `borg.crypto.low_level.AES256_CTR_HMAC_SHA256` - it's not that hard

Reply:
ok, incremental improvement. \o/ and we can kill the AES class once nobody is using v1 keys any more.